util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,238 @@
+        json.dump(resources, f, indent=4)
+
+
+def progress_hook(t):
+    last_b = [0]
+
+    def update_to(b=1, bsize=1, tsize=None):
+        if tsize not in (None, -1):
+            t.total = tsize
+        displayed = t.update((b - last_b[0]) * bsize)
+        last_b[0] = b
+        return displayed
+
+    return update_to
+
+
+def download_resources(output: pathlib.Path):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    with open(output.joinpath("resources.json")) as f:
+        resources = json.load(f)
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filename = pathlib.Path(
+                *pathlib.Path(parse.urlsplit(url).path).parts[3:]
+            )
+            filepath = pathlib.Path(path).joinpath(filename)
+            filepath.parent.mkdir(parents=True, exist_ok=True)
+
+            if (
+                not filepath.exists()
+                or hashlib.md5(filepath.read_bytes()).hexdigest()
+                != resource["md5sum"]
+            ):
+                with tqdm(
+                    unit="B",
+                    unit_scale=True,
+                    unit_divisor=1024,
+                    miniters=1,
+                    desc=f"Downloading {url}",
+                ) as t:
+                    request.urlretrieve(url, filepath, progress_hook(t))
+            resource["url"] = filepath.absolute().as_uri()
+
+    with open(output.joinpath("resources.json"), "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+if __name__ == "__main__" or __name__ == "__m5_main__":
+    parser = argparse.ArgumentParser(
+        description="Get resources from the database"
+    )
+
+    parser.add_argument(
+        "--config-file-url",
+        type=str,
+        default="https://raw.githubusercontent.com/gem5/gem5/stable/"
+        "src/python/gem5_default_config.py",
+        help="URL to the gem5 config file",
+    )
+
+    parser.add_argument(
+        "--output_dir",
+        type=str,
+        default=pathlib.Path.cwd(),
+        help="Output directory absolute path, default is the cwd."
+        "The resources.json and all resources will be saved in this directory",
+    )
+    args = parser.parse_args()
+
+    output_path = pathlib.Path(args.output_dir)
+
+    config_file_suffix = pathlib.Path(
+        parse.urlsplit(args.config_file_url).path
+    ).suffix
+
+    # Get the gem5 config file from the url
+    if config_file_suffix == ".py":
+        gem5_config = request.urlopen(args.config_file_url)
+        gem5_config = gem5_config.read().decode("utf-8").split("=")[-1]
+        gem5_config = eval(gem5_config)
+        gem5_config = gem5_config["sources"]["gem5-resources"]
```

Comment:
So I can understand this can we improve this by having each line set a new variable, each of which describes it's value. E.g., first can be `request_returned`, the secon can be `decoded_request` etc. 

Also, please explain what `.split("="][-1]` is doing.